### PR TITLE
[GOBBLIN-319] Add DatasetResolver to transform raw Gobblin dataset to application specific dataset

### PR DESCRIPTION
**Pull request by zxcware**
Dear Gobblin maintainers,

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title.
    - https://issues.apache.org/jira/browse/GOBBLIN-319

### Description
- [x] Here are some details about my PR:
  - Add `DatasetResolver` to transform raw Gobblin dataset to application specific dataset
  - Fix lineage info not set while publishing single task data in `BaseDataPublisher`

### Tests
- [x] My PR adds the following unit tests:
  - `LineageEventTest#testEvent`
  - `BaseDataPublisherTest#testPublishSingleTask`
  - `BaseDataPublisherTest#testPublishMultiTasks`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by zxcware** [comment]
@yukuai518 Please review

File: gobblin-core/src/test/java/org/apache/gobblin/publisher/BaseDataPublisherTest.java [hunk 1]
```diff
@@ -588,4 +621,23 @@ private State buildDefaultState(int numBranches)
...
+    state.setProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY, "namespace");
+    state.setProp(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY, "table");
+    state.setProp(ConfigurationKeys.WRITER_FILE_PATH_TYPE, "namespace_table");
+    state.setProp(ConfigurationKeys.FORK_BRANCHES_KEY, numBranches);
+    state.setProp(ConfigurationKeys.DATA_PUBLISHER_FINAL_DIR, "/data/output");
+    state.setProp(ConfigurationKeys.WRITER_OUTPUT_DIR, "/data/working");
+    if (numBranches > 1) {
+      for (int i = 0; i < numBranches; i++) {
+        state.setProp(ConfigurationKeys.DATA_PUBLISHER_FINAL_DIR + "." + i, "/data/output" + "/branch" + i);
+        state.setProp(ConfigurationKeys.WRITER_OUTPUT_DIR+ "." + i, "/data/working" + "/branch" + i);
```
Comment (shuzhang1989) [inline gobblin-core/src/test/java/org/apache/gobblin/publisher/BaseDataPublisherTest.java]: nit: space between WRITER_OUTPUT_DIR and  "."

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 2]
```diff
@@ -210,6 +210,7 @@ private void maySubmitLineageEvent(JobState.DatasetState datasetState) {
       }
     }
     if (states.size() == 0) {
+      log.info("Will not submit lineage events");
```
Comment (shuzhang1989) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: maybe change it to include the reason why? 
"will not submit lineage events due to empty tasks"

File: gobblin-core/src/test/java/org/apache/gobblin/publisher/BaseDataPublisherTest.java (hunk 1 above)
Comment (zxcware) [reply gobblin-core/src/test/java/org/apache/gobblin/publisher/BaseDataPublisherTest.java]: +1

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java (hunk 2 above)
Comment (zxcware) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: +1

File: gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java [hunk 3]
```diff
@@ -77,8 +81,10 @@ private LineageInfo() {
    *
    */
   public static void setSource(DatasetDescriptor source, State state) {
-    state.setProp(getKey(NAME_KEY), source.getName());
-    state.setProp(getKey(LineageEventBuilder.SOURCE), GSON.toJson(source));
+    DatasetResolver resolver = getResolver(state);
```
Comment (ibuenros) [inline gobblin-metrics-libs/gobblin-metrics-base/src/main/java/org/apache/gobblin/metrics/event/lineage/LineageInfo.java]: Are you re-instantiating the dataset resolver for each work unit?

File: gobblin-utility/src/main/java/org/apache/gobblin/util/FileListUtils.java [hunk 4]
```diff
@@ -56,6 +57,32 @@ public boolean accept(Path path) {
     }
   };
 
+  /**
+   * Get any data file, which is not a directory, from the given path
+   */
+  public static FileStatus getAnyFile(FileSystem fs, Path path) throws IOException {
```
Comment (ibuenros) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/FileListUtils.java]: Is this used anywhere?

**Pull request closed without merging**